File: robot_inference/scripts/robot_inference_client.py
#!/usr/bin/env python

import sys
import rospy
from robot_inference.srv import *
import array

reverse_table = {0:"stay", 4:"right", 3:"left", 1:"up", 2:"down"}
table = {"stay":0, "right":4, "left":3, "up":1, "down":2}
def translate(i):
    if i in table:
        return table[i]
    return i
def restore(i):
    return reverse_table[i]
def demonstrate(inference):
    lst = list(eval(str(inference)))
    sequence = []
    for i in range(0,len(lst)-2,3):
        action = restore(lst[i+2])
        sequence.append((lst[i],lst[i+1],action))
    for i in sequence:
        print(i)
def load_observations(filename):
    # loads a list of observations saved by save_observations()
    f = open(filename, 'r')
    print ("Filename: "+filename)
    observations = []
    for line in f.readlines():
        line = line.strip()
        if line == 'missing':
            observations.append(None)
        else:
            parts  = line.split()
            x      = int(parts[0])
            y      = int(parts[1])
            observations.append( (x, y) )

    f.close()
    result = []
    for state in observations:
        for entry in state:
            result.append(translate(entry))
    return result

def robot_inference_client(x):
    rospy.wait_for_service('inference')
    try:
        infer = rospy.ServiceProxy('inference', Inference)
        resp1 = infer(x)
        return resp1.inference
    except rospy.ServiceException, e:
        print ("Service call failed: %s"%e)

def usage():
    return "infer robot behavior from observations"

if __name__ == "__main__":
    print ("file name of observations (put in the same folder)")
    if len(sys.argv) >= 2:
        filename = sys.argv[1]
	x = load_observations(filename)
    else:
        print (usage())
        sys.exit(1)
    print ("Requesting")
    print ("Infered behavior:")
    inf = robot_inference_client(x)
    demonstrate(inf)

File: robot_inference/src/robot_inference_client.cpp
#include "ros/ros.h"
#include "robot_inference/Inference.h"
#include <cstdlib>
#include <string>
#include <iostream>
#include <fstream>
#include <sstream>

using namespace std;

vector<int> load_observations(string filename)
{
  vector<int> observations;
  ifstream fin;
  fin.open(filename.c_str());
  if (fin.fail())
  {
    throw invalid_argument("fail to open the file");
  }
  for (int i = 0; fin.good(); i++)
  {
    stringstream linestream;
    string line;
    getline(fin, line);
    // skip meanlingless lines 
    if (line[0] == ' ' || line == "")
    {
      i--;
      continue;
    }
    linestream << line;
    int x,y;
    linestream >> x >> y;
    observations.push_back(x);
    observations.push_back(y);
    linestream.clear();
  }
  fin.close();
  return observations;
}

string mapping(int val)
{ 
  if (val == 4) return "right";
  else if (val == 3) return "left";
  else if (val == 1) return "up";
  else if (val == 2) return "down";
  else return "stay";
}

// define my own function to convert from int to string in ros standard c++03
string convert(int num)
{
  ostringstream convert;
  convert << num;
  return convert.str();
}

int main(int argc, char *argv[])
{
  ROS_INFO("file name of observations (put in the same folder)");
  ros::init(argc, argv, "robot_inference_client");
  if (argc < 2)
  {
    ROS_INFO("infer robot behavior from observations");
    return 1;
  }

  ros::NodeHandle n;
  ros::ServiceClient client = n.serviceClient<robot_inference::Inference>("inference");
  robot_inference::Inference srv;
  string filename = argv[1];
  vector<int> observations;
  try
  {
    observations = load_observations("test.txt");
  }
  catch(const invalid_argument& e)
  {
    ROS_ERROR("Invalid observation file");
    return 1;
  } 
  srv.request.observation = observations;

  if (client.call(srv))
  {
    vector<int> inference = srv.response.inference;
    string display = "\n";
    for (int i = 0; i < inference.size() / 3; i++)
    {
      display += "(" + convert(inference[3 * i]) + ", " + convert(inference[3 * i + 1]) + "), ";
      display += mapping(inference[3 * i + 2]) + "\n";
    }
    ROS_INFO("%s", display.c_str());
  }
  else
  {
    ROS_ERROR("Failed to call service robot_inference");
    return 1;
  }

  return 0;
}

